### PR TITLE
Add a tool `sdk-lockfiles` to audit lockfiles

### PR DESCRIPTION
**Pull request by ysaito1001**
## Motivation and Context
See [README](https://github.com/smithy-lang/smithy-rs/blob/b9f363352603a14305a7cf7ca0317eaf431f638e/tools/ci-build/sdk-lockfiles/README.md)

A later PR will use this tool to enforce the check (likely during the execution of pre-commit-hooks).

## Testing
- Added unit tests for the tool
- Ran the tool against currently checked-in lockflies in the `main` branch and ensured the SDK lockfile contained dependencies listed in the other runtime lockfiles.

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

**Review by aajtodd — APPROVED** [review]

File: tools/ci-build/sdk-lockfiles/src/audit.rs [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+];
+
+fn new_dependency_for_aws_sdk(crate_name: &str) -> bool {
+    AWS_SDK_RUNTIMES.contains(&crate_name)
+        || crate_name == "inlineable"
+        || (crate_name.starts_with("aws-smithy-")
+            && !SERVER_SPECIFIC_RUNTIMES.contains(&crate_name))
+}
+
+// Recursively traverses a chain of dependencies originating from a potential new dependency. Returns true as soon as
```
Comment (drganjoo) [inline tools/ci-build/sdk-lockfiles/src/audit.rs:51]: Nit: The word dependency is a bit confusing as it indicates that we will look for the dependencies of the provided node, where as we are looking for crates that consume this dependency.

How about:

// This function checks if the given dependency is directly or indirectly consumed by any AWS SDK runtime crate.
// It recursively traverses the chain of consumers starting from the given Cargo dependency.
// Returns true as soon as it encounters a crate name that matches one of the AWS SDK runtime crates.

File: tools/ci-build/sdk-lockfiles/src/audit.rs [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+// it encounters a crate name that matches a runtime crate used by the AWS SDK.
+fn visit(graph: &Graph, node_index: NodeIndex, visited: &mut BTreeSet<NodeIndex>) -> bool {
+    if !visited.insert(node_index) {
+        return false;
+    }
+
+    let dependencies = graph
```
Comment (drganjoo) [inline tools/ci-build/sdk-lockfiles/src/audit.rs:58]: Nit: how about renaming this to `consumers`? This makes it clear that we are not looking at further dependencies of the given node but are interested in the crates that depend on the `node_index`.

**Review by drganjoo — APPROVED** [review]

File: tools/ci-build/sdk-lockfiles/src/audit.rs [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+// https://github.com/smithy-lang/smithy-rs/blob/0f9b9aba386ea3063912a0464ba6a1fd7c596018/buildSrc/src/main/kotlin/CrateSet.kt#L85-L87
+const SERVER_SPECIFIC_RUNTIMES: &[&str] = &[
+    "aws-smithy-http-server",
+    "aws-smithy-http-server-python",
+    "aws-smithy-http-typescript",
+];
+
+fn new_dependency_for_aws_sdk(crate_name: &str) -> bool {
+    AWS_SDK_RUNTIMES.contains(&crate_name)
+        || crate_name == "inlineable"
```
[the remark this answers is not in the document]
Comment (ysaito1001) [reply tools/ci-build/sdk-lockfiles/src/audit.rs:46]: Depth first traversal for dependency check is performed in runtime lockfiles, importantly `rust-runtime/Cargo.lock` being one of them. The crate `inlineable` is [listed in that lockfile](https://github.com/smithy-lang/smithy-rs/blob/main/rust-runtime/Cargo.lock#L1998-L2024). This means that `inlineable`'s [dependencies](https://github.com/smithy-lang/smithy-rs/blob/main/rust-runtime/Cargo.lock#L2001-L2024) will be part of a set of crates used by a code-generated client SDK. That further means if an `inlineable`'s dependency were not listed in the SDK lockfile, compiling that client SDK crate would fetch the latest version of that unpinned crate from crates.io. We know that unpinned crate with its version subject to change is a problem for reproducible build.

This is the reason the `inlineable` crate also needs to be checked. There is room for improvement, though. A new dependency introduced in `inlineable` could only be used by the server SDK, in which case, that dependency is not needed by the client SDK, and the SDK lockfile doesn't need to include it. But this depth first traversal for dependency check is not smart enough to handle that case. We err on the side of caution by always considering a dependency introduced to the `inlineable` as required by the code-generated client SDK.

File: tools/ci-build/sdk-lockfiles/README.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+=============
+
+This CLI tool audits the `Cargo.lock` files in the `smithy-rs` repository. These lockfiles are used to ensure
+reproducible builds. The `sdk-lockfiles` tool specifically audits the following lockfiles:
+- The [lockfile](https://github.com/smithy-lang/smithy-rs/blob/main/rust-runtime/Cargo.lock) for Smithy runtime crates
+- The [lockfile](https://github.com/smithy-lang/smithy-rs/blob/main/aws/rust-runtime/Cargo.lock) for AWS runtime crates
+- The [lockfile](https://github.com/smithy-lang/smithy-rs/blob/main/aws/rust-runtime/aws-config/Cargo.lock) for the `aws-config` crate
+- The [lockfile](https://github.com/smithy-lang/smithy-rs/blob/main/aws/sdk/Cargo.lock) for the workspace containing code-generated AWS SDK crates (*)
+
+Specifically, the tool ensures that the lockfile marked with (*) is a superset containing all dependencies listed in
```
[the remark this answers is not in the document]
Comment (ysaito1001) [reply tools/ci-build/sdk-lockfiles/README.md:11]: > What happens if different versions of the same crate are used across the first three lockfiles?

The first three lockfiles, e.g., runtime lockfiles, are used for building their workspaces deterministically, and they are isolated to one another. So there is no problem having different versions of the same crate. When runtime crates are "assembled" to a generated client SDK via `./gradlew aws:sdk:assemble`, the first three lockfiles are no longer relevant; when this client SDK is built, it'll instead refer to the SDK lockfile

File: tools/ci-build/sdk-lockfiles/src/audit.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ysaito1001) [reply tools/ci-build/sdk-lockfiles/src/audit.rs:46]: Good question. Depth first traversal for dependency check is performed in runtime lockfiles, importantly `rust-runtime/Cargo.lock` being one of them. The crate `inlineable` is [listed in that lockfile](https://github.com/smithy-lang/smithy-rs/blob/main/rust-runtime/Cargo.lock#L1998-L2024). This means that `inlineable`'s [dependencies](https://github.com/smithy-lang/smithy-rs/blob/main/rust-runtime/Cargo.lock#L2001-L2024) will be part of a set of crates used by a code-generated client SDK, but _only when_ they are referred to by [CargoDependency.kt](https://github.com/smithy-lang/smithy-rs/blob/main/codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/rustlang/CargoDependency.kt).

In other words, I think your question just exposed a hole in the current audit mechanism in `sdk-lockfiles`. `sdk-lockfiles` (or maybe somewhere outside the tool, in which case we need to call that out as a tool's limitation) needs to check whether the SDK lockfile is behind `CargoDependency.kt` or not; if it's behind, audit should fail.

And we don't need to track `aws-inlineable` or `inlineable` here because knowing whether the SDK lockfile is behind `CargoDependency.kt` or not isn't possible as part of depth first traversal.

File: tools/ci-build/sdk-lockfiles/src/audit.rs [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+        .iter()
+        .map(|dep| {
+            let package = lockfile
+                .packages
+                .iter()
+                .find(|pkg| pkg.name.as_str() == dep)
+                .unwrap();
+            tree.nodes()[&package.into()]
+        })
+        .collect();
```
[the remark this answers is not in the document]
Comment (ysaito1001) [reply tools/ci-build/sdk-lockfiles/src/audit.rs:101]: Yeah, that makes sense, given `[target.to_owned()]` contains a single element

File: tools/ci-build/sdk-lockfiles/src/audit.rs (hunk 1 above)
Comment (ysaito1001) [reply tools/ci-build/sdk-lockfiles/src/audit.rs:51]: I think the suggested comment is more for a comment for the `new_dependency` function. This comment `Recursively traverses...` is for the helper used by `new_dependency`.

File: tools/ci-build/sdk-lockfiles/src/audit.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ysaito1001) [reply tools/ci-build/sdk-lockfiles/src/audit.rs:46]: Good question. Depth first traversal for dependency check is performed in runtime lockfiles, importantly `rust-runtime/Cargo.lock` being one of them. The crate `inlineable` is [listed in that lockfile](https://github.com/smithy-lang/smithy-rs/blob/main/rust-runtime/Cargo.lock#L1998-L2024). This means that `inlineable`'s [dependencies](https://github.com/smithy-lang/smithy-rs/blob/main/rust-runtime/Cargo.lock#L2001-L2024) will be part of a set of crates used by a code-generated client SDK, but _only when_ they are referred to by [CargoDependency.kt](https://github.com/smithy-lang/smithy-rs/blob/main/codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/rustlang/CargoDependency.kt).

We don't need to track `aws-inlineable` or `inlineable` here because knowing whether the SDK lockfile is behind `CargoDependency.kt` or not isn't possible as part of depth first traversal. That should probably be done by a parent script using this tool during the release process.
